File: examples/rust/multi_curve_calibration_example.rs
//! Multi-curve yield curve calibration example.
//!
//! Demonstrates post-2008 multi-curve framework with:
//! - OIS discount curve for discounting
//! - Separate forward curves per tenor (1M, 3M, 6M)

use finstack::calibration::{
    bootstrap::{DiscountCurveCalibrator, ForwardCurveCalibrator},
    quote::{FutureSpecs, RatesQuote},
    CalibrationConfig, Calibrator,
};
use finstack::currency::Currency;
use finstack::dates::{Date, DayCount, Frequency};
use finstack::market_data::{
    context::MarketContext,
    interp::types::InterpStyle,
    traits::{Discount, Forward, TermStructure},
};
use finstack::prelude::*;
use time::Month;

fn main() -> finstack::Result<()> {
    let base_date = Date::from_calendar_date(2025, Month::January, 1)?;
    
    // Step 1: Create market context
    let mut context = MarketContext::new();
    
    // Step 2: Calibrate OIS discount curve
    println!("Calibrating OIS discount curve...");
    let ois_quotes = vec![
        // Overnight deposit
        RatesQuote::Deposit {
            maturity: base_date + time::Duration::days(1),
            rate: 0.0450,
            day_count: DayCount::Act365F,
        },
        // OIS swaps
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(30),
            rate: 0.0452,
            fixed_freq: Frequency::annual(),
            float_freq: Frequency::daily(),
            fixed_dc: DayCount::Act365F,
            float_dc: DayCount::Act365F,
            index: "USD-OIS".to_string(),
        },
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(90),
            rate: 0.0455,
            fixed_freq: Frequency::annual(),
            float_freq: Frequency::daily(),
            fixed_dc: DayCount::Act365F,
            float_dc: DayCount::Act365F,
            index: "USD-OIS".to_string(),
        },
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(180),
            rate: 0.0458,
            fixed_freq: Frequency::annual(),
            float_freq: Frequency::daily(),
            fixed_dc: DayCount::Act365F,
            float_dc: DayCount::Act365F,
            index: "USD-OIS".to_string(),
        },
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(365),
            rate: 0.0462,
            fixed_freq: Frequency::annual(),
            float_freq: Frequency::daily(),
            fixed_dc: DayCount::Act365F,
            float_dc: DayCount::Act365F,
            index: "USD-OIS".to_string(),
        },
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(365 * 2),
            rate: 0.0468,
            fixed_freq: Frequency::annual(),
            float_freq: Frequency::daily(),
            fixed_dc: DayCount::Act365F,
            float_dc: DayCount::Act365F,
            index: "USD-OIS".to_string(),
        },
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(365 * 5),
            rate: 0.0475,
            fixed_freq: Frequency::annual(),
            float_freq: Frequency::daily(),
            fixed_dc: DayCount::Act365F,
            float_dc: DayCount::Act365F,
            index: "USD-OIS".to_string(),
        },
    ];
    
    let ois_calibrator = DiscountCurveCalibrator::new(
        "USD-OIS-DISC",
        base_date,
        Currency::USD,
    )
    .with_solve_interp(InterpStyle::MonotoneConvex)
    .with_config(CalibrationConfig {
        tolerance: 1e-10,
        max_iterations: 100,
        ..Default::default()
    });
    
    let (ois_curve, ois_report) = ois_calibrator.calibrate(&ois_quotes, &context)?;
    
    println!("OIS curve calibrated:");
    println!("  - Success: {}", ois_report.success);
    println!("  - Iterations: {}", ois_report.iterations);
    println!("  - Max residual: {:.2e}", ois_report.max_residual);
    println!("  - RMSE: {:.2e}", ois_report.rmse);
    
    // Update context with OIS curve and set collateral mapping
    context = context
        .insert_discount(ois_curve)
        .map_collateral("USD-CSA", "USD-OIS-DISC".into());
    
    // Step 3: Calibrate 3M forward curve
    println!("\nCalibrating 3M SOFR forward curve...");
    let forward_3m_quotes = vec![
        // FRAs
        RatesQuote::FRA {
            start: base_date + time::Duration::days(30),
            end: base_date + time::Duration::days(120),
            rate: 0.0463,
            day_count: DayCount::Act360,
        },
        RatesQuote::FRA {
            start: base_date + time::Duration::days(90),
            end: base_date + time::Duration::days(180),
            rate: 0.0465,
            day_count: DayCount::Act360,
        },
        RatesQuote::FRA {
            start: base_date + time::Duration::days(180),
            end: base_date + time::Duration::days(270),
            rate: 0.0468,
            day_count: DayCount::Act360,
        },
        // Futures
        RatesQuote::Future {
            expiry: base_date + time::Duration::days(90),
            price: 95.30, // Implies 4.70% rate
            specs: FutureSpecs {
                multiplier: 2500.0,
                face_value: 1_000_000.0,
                delivery_months: 3,
                day_count: DayCount::Act360,
                convexity_adjustment: Some(0.0002), // 2bp convexity adjustment
            },
        },
        // Vanilla IRS vs 3M SOFR
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(365),
            rate: 0.0470,
            fixed_freq: Frequency::semi_annual(),
            float_freq: Frequency::quarterly(),
            fixed_dc: DayCount::Thirty360,
            float_dc: DayCount::Act360,
            index: "USD-SOFR-3M".to_string(),
        },
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(365 * 2),
            rate: 0.0475,
            fixed_freq: Frequency::semi_annual(),
            float_freq: Frequency::quarterly(),
            fixed_dc: DayCount::Thirty360,
            float_dc: DayCount::Act360,
            index: "USD-SOFR-3M".to_string(),
        },
        RatesQuote::Swap {
            maturity: base_date + time::Duration::days(365 * 5),
            rate: 0.0482,
            fixed_freq: Frequency::semi_annual(),
            float_freq: Frequency::quarterly(),
            fixed_dc: DayCount::Thirty360,
            float_dc: DayCount::Act360,
            index: "USD-SOFR-3M".to_string(),
        },
    ];
    
    let forward_3m_calibrator = ForwardCurveCalibrator::new(
        "USD-SOFR-3M-FWD",
        0.25, // 3 months = 0.25 years
        base_date,
        Currency::USD,
        "USD-OIS-DISC",
    )
    .with_solve_interp(InterpStyle::Linear)
    .with_config(CalibrationConfig {
        tolerance: 1e-10,
        max_iterations: 100,
        ..Default::default()
    });
    
    let (forward_3m_curve, forward_3m_report) = 
        forward_3m_calibrator.calibrate(&forward_3m_quotes, &context)?;
    
    println!("3M forward curve calibrated:");
    println!("  - Success: {}", forward_3m_report.success);
    println!("  - Iterations: {}", forward_3m_report.iterations);
    println!("  - Max residual: {:.2e}", forward_3m_report.max_residual);
    println!("  - RMSE: {:.2e}", forward_3m_report.rmse);
    
    context = context.insert_forward(forward_3m_curve);
    
    // Step 4: Display some curve values
    println!("\n=== Curve Values ===");
    
    let disc_curve = context.disc("USD-OIS-DISC")?;
    let fwd_3m_curve = context.fwd("USD-SOFR-3M-FWD")?;
    
    println!("\nOIS Discount Factors:");
    for t in &[0.25, 0.5, 1.0, 2.0, 5.0] {
        println!("  t = {:.2} years: DF = {:.6}", t, disc_curve.df(*t));
    }
    
    println!("\n3M Forward Rates:");
    for t in &[0.0, 0.25, 0.5, 1.0, 2.0, 5.0] {
        println!("  t = {:.2} years: Fwd = {:.2}%", t, fwd_3m_curve.rate(*t) * 100.0);
    }
    
    // Step 5: Demonstrate basis spread (simplified)
    println!("\n=== Basis Spreads ===");
    println!("In a full implementation, we would calibrate 1M and 6M curves");
    println!("and use basis swaps to ensure consistency between tenors.");
    
    // Example basis swap quote (not processed in this simple example)
    let _basis_swap = RatesQuote::BasisSwap {
        maturity: base_date + time::Duration::days(365 * 2),
        primary_index: "USD-SOFR-3M".to_string(),
        reference_index: "USD-SOFR-6M".to_string(),
        spread_bp: 2.5, // 3M pays 6M + 2.5bp
        primary_freq: Frequency::quarterly(),
        reference_freq: Frequency::semi_annual(),
        primary_dc: DayCount::Act360,
        reference_dc: DayCount::Act360,
        currency: Currency::USD,
    };
    
    println!("\nMulti-curve calibration complete!");
    println!("Market context contains:");
    println!("  - Discount curve: USD-OIS-DISC");
    println!("  - Forward curve: USD-SOFR-3M-FWD");
    println!("  - Collateral mapping: USD-CSA -> USD-OIS-DISC");
    
    Ok(())
}

File: examples/python/valuations/multi_curve_calibration_example.py
#!/usr/bin/env python3
"""
Multi-curve yield curve calibration example.

Demonstrates post-2008 multi-curve framework with:
- OIS discount curve for discounting  
- Separate forward curves per tenor (1M, 3M, 6M)
- Market standard conventions
"""

import finstack as fs
from datetime import date, timedelta

def main():
    base_date = date(2025, 1, 1)
    
    print("Multi-Curve Calibration Example")
    print("=" * 50)
    
    # Step 1: Calibrate OIS discount curve
    print("\n1. Calibrating OIS discount curve...")
    
    # OIS quotes (deposits + OIS swaps)
    ois_quotes = [
        # Overnight deposit
        fs.RatesQuote.deposit(
            maturity=base_date + timedelta(days=1),
            rate=0.0450,
            day_count=fs.DayCount.Act365F
        ),
        # OIS swaps
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=30),
            rate=0.0452,
            fixed_freq=fs.Frequency.annual(),
            float_freq=fs.Frequency.daily(),
            fixed_dc=fs.DayCount.Act365F,
            float_dc=fs.DayCount.Act365F,
            index="USD-OIS"
        ),
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=90),
            rate=0.0455,
            fixed_freq=fs.Frequency.annual(),
            float_freq=fs.Frequency.daily(),
            fixed_dc=fs.DayCount.Act365F,
            float_dc=fs.DayCount.Act365F,
            index="USD-OIS"
        ),
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=180),
            rate=0.0458,
            fixed_freq=fs.Frequency.annual(),
            float_freq=fs.Frequency.daily(),
            fixed_dc=fs.DayCount.Act365F,
            float_dc=fs.DayCount.Act365F,
            index="USD-OIS"
        ),
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=365),
            rate=0.0462,
            fixed_freq=fs.Frequency.annual(),
            float_freq=fs.Frequency.daily(),
            fixed_dc=fs.DayCount.Act365F,
            float_dc=fs.DayCount.Act365F,
            index="USD-OIS"
        ),
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=365*2),
            rate=0.0468,
            fixed_freq=fs.Frequency.annual(),
            float_freq=fs.Frequency.daily(),
            fixed_dc=fs.DayCount.Act365F,
            float_dc=fs.DayCount.Act365F,
            index="USD-OIS"
        ),
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=365*5),
            rate=0.0475,
            fixed_freq=fs.Frequency.annual(),
            float_freq=fs.Frequency.daily(),
            fixed_dc=fs.DayCount.Act365F,
            float_dc=fs.DayCount.Act365F,
            index="USD-OIS"
        ),
    ]
    
    # Create calibrator with Hagan-West monotone convex interpolation
    ois_calibrator = fs.DiscountCurveCalibrator(
        curve_id="USD-OIS-DISC",
        base_date=base_date,
        currency=fs.Currency.USD
    ).with_solve_interp(fs.InterpStyle.MonotoneConvex)
    
    # Calibrate OIS curve
    context = fs.MarketContext()
    ois_curve, ois_report = ois_calibrator.calibrate(ois_quotes, context)
    
    print(f"  Success: {ois_report.success}")
    print(f"  Iterations: {ois_report.iterations}")
    print(f"  Max residual: {ois_report.max_residual:.2e}")
    print(f"  RMSE: {ois_report.rmse:.2e}")
    
    # Update context with OIS curve and collateral mapping
    context = context.insert_discount(ois_curve).map_collateral("USD-CSA", "USD-OIS-DISC")
    
    # Step 2: Calibrate 3M forward curve
    print("\n2. Calibrating 3M SOFR forward curve...")
    
    # 3M forward quotes (FRAs + futures + vanilla IRS)
    forward_3m_quotes = [
        # FRAs
        fs.RatesQuote.fra(
            start=base_date + timedelta(days=30),
            end=base_date + timedelta(days=120),
            rate=0.0463,
            day_count=fs.DayCount.Act360
        ),
        fs.RatesQuote.fra(
            start=base_date + timedelta(days=90),
            end=base_date + timedelta(days=180),
            rate=0.0465,
            day_count=fs.DayCount.Act360
        ),
        fs.RatesQuote.fra(
            start=base_date + timedelta(days=180),
            end=base_date + timedelta(days=270),
            rate=0.0468,
            day_count=fs.DayCount.Act360
        ),
        # Futures
        fs.RatesQuote.future(
            expiry=base_date + timedelta(days=90),
            price=95.30,  # Implies 4.70% rate
            specs=fs.FutureSpecs(
                multiplier=2500.0,
                face_value=1_000_000.0,
                delivery_months=3,
                day_count=fs.DayCount.Act360,
                convexity_adjustment=0.0002  # 2bp convexity adjustment
            )
        ),
        # Vanilla IRS vs 3M SOFR
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=365),
            rate=0.0470,
            fixed_freq=fs.Frequency.semi_annual(),
            float_freq=fs.Frequency.quarterly(),
            fixed_dc=fs.DayCount.Thirty360,
            float_dc=fs.DayCount.Act360,
            index="USD-SOFR-3M"
        ),
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=365*2),
            rate=0.0475,
            fixed_freq=fs.Frequency.semi_annual(),
            float_freq=fs.Frequency.quarterly(),
            fixed_dc=fs.DayCount.Thirty360,
            float_dc=fs.DayCount.Act360,
            index="USD-SOFR-3M"
        ),
        fs.RatesQuote.swap(
            maturity=base_date + timedelta(days=365*5),
            rate=0.0482,
            fixed_freq=fs.Frequency.semi_annual(),
            float_freq=fs.Frequency.quarterly(),
            fixed_dc=fs.DayCount.Thirty360,
            float_dc=fs.DayCount.Act360,
            index="USD-SOFR-3M"
        ),
    ]
    
    # Create forward curve calibrator
    forward_3m_calibrator = fs.ForwardCurveCalibrator(
        fwd_curve_id="USD-SOFR-3M-FWD",
        tenor_years=0.25,  # 3 months = 0.25 years
        base_date=base_date,
        currency=fs.Currency.USD,
        discount_curve_id="USD-OIS-DISC"
    ).with_solve_interp(fs.InterpStyle.Linear)
    
    # Calibrate forward curve
    forward_3m_curve, forward_3m_report = forward_3m_calibrator.calibrate(
        forward_3m_quotes, context
    )
    
    print(f"  Success: {forward_3m_report.success}")
    print(f"  Iterations: {forward_3m_report.iterations}")
    print(f"  Max residual: {forward_3m_report.max_residual:.2e}")
    print(f"  RMSE: {forward_3m_report.rmse:.2e}")
    
    context = context.insert_forward(forward_3m_curve)
    
    # Step 3: Display curve values
    print("\n3. Curve Values")
    print("-" * 30)
    
    disc_curve = context.disc("USD-OIS-DISC")
    fwd_3m_curve = context.fwd("USD-SOFR-3M-FWD")
    
    print("\nOIS Discount Factors:")
    for t in [0.25, 0.5, 1.0, 2.0, 5.0]:
        print(f"  t = {t:4.2f} years: DF = {disc_curve.df(t):.6f}")
    
    print("\n3M Forward Rates:")
    for t in [0.0, 0.25, 0.5, 1.0, 2.0, 5.0]:
        print(f"  t = {t:4.2f} years: Fwd = {fwd_3m_curve.rate(t)*100:.2f}%")
    
    # Step 4: Demonstrate basis spreads (conceptual)
    print("\n4. Basis Spreads")
    print("-" * 30)
    print("In a full implementation, we would:")
    print("  - Calibrate 1M and 6M forward curves")
    print("  - Use basis swaps to ensure tenor consistency")
    print("  - Apply basis adjustments to align curves")
    
    # Example basis swap quote (for illustration)
    basis_swap = fs.RatesQuote.basis_swap(
        maturity=base_date + timedelta(days=365*2),
        primary_index="USD-SOFR-3M",
        reference_index="USD-SOFR-6M",
        spread_bp=2.5,  # 3M pays 6M + 2.5bp
        primary_freq=fs.Frequency.quarterly(),
        reference_freq=fs.Frequency.semi_annual(),
        primary_dc=fs.DayCount.Act360,
        reference_dc=fs.DayCount.Act360,
        currency=fs.Currency.USD
    )
    print(f"\nExample basis swap: 3M vs 6M spread = {basis_swap.spread_bp}bp")
    
    # Step 5: Simple calibration alternative
    print("\n5. Alternative: SimpleCalibration")
    print("-" * 30)
    print("For convenience, use SimpleCalibration to calibrate all curves at once:")
    
    # Combine all quotes
    all_quotes = [fs.MarketQuote.rates(q) for q in ois_quotes + forward_3m_quotes]
    
    # Use SimpleCalibration for one-shot calibration
    simple_cal = fs.SimpleCalibration(base_date, fs.Currency.USD)
    final_context, final_report = simple_cal.calibrate(all_quotes)
    
    print(f"  Total iterations: {final_report.iterations}")
    print(f"  Success: {final_report.success}")
    
    print("\n" + "=" * 50)
    print("Multi-curve calibration complete!")
    print("\nMarket context contains:")
    print("  - Discount curve: USD-OIS-DISC")
    print("  - Forward curve: USD-SOFR-3M-FWD")
    print("  - Collateral mapping: USD-CSA -> USD-OIS-DISC")

if __name__ == "__main__":
    main()
